### PR TITLE
skip binary replacement on win; force / in replacement prefix on win.

### PR DESCRIPTION
**Pull request by msarahan**
This might be a fix for https://github.com/conda/conda-build/issues/997

### CODE REVIEW

**Comment by kalefranz** [comment]
@ilanschnell Any objections?

**Comment by kalefranz** [comment]
The most recent commit here contains regression tests for #2626

**Comment by msarahan** [comment]
@ilanschnell and I talked a fair bit about this.  We settled on this solution together.  I'll let him have his say here, though.

**Comment by ilanschnell** [comment]
Looks good to me.

**Comment by kalefranz** [comment]
____________________________ FileTests.test_binary ____________________________
    Traceback (most recent call last):
      File "C:\projects\conda\tests\test_install.py", line 106, in test_binary
        placeholder='/some-placeholder', mode='binary')
      File "C:\projects\conda\conda\install.py", line 391, in update_prefix
        data = replace_prefix(mode, data, placeholder, new_prefix)
      File "C:\projects\conda\conda\install.py", line 377, in replace_prefix
        sys.exit("Invalid mode:" % mode)
    TypeError: not all arguments converted during string formatting
     6 failed, 217 passed, 2 skipped, 2 xfailed, 1 pytest-warnings in 734.32 seconds

File: conda/install.py [hunk 1]
```diff
@@ -369,16 +369,18 @@ def replace_long_shebang(mode, data):
 def replace_prefix(mode, data, placeholder, new_prefix):
     if mode == 'text':
         data = data.replace(placeholder.encode('utf-8'), new_prefix.encode('utf-8'))
-    elif mode == 'binary':
+    # Skip binary replacement in Windows.  Some files do have prefix information embedded, but
+    #    this should not matter, as it is not used for things like RPATH.
+    elif mode == 'binary' and sys.platform != "win32":
```
Comment (kalefranz) [inline conda/install.py]: We should use one of the five `on_win` functions here instead of `sys.platform` directly.

**Comment by msarahan** [comment]
Hooray for testing! I'll fix this first thing tomorrow.

File: conda/install.py (hunk 1 above)
Comment (msarahan) [inline conda/install.py]: Thanks, good call. I have gotten lazy about checking whether on_win is defined.

File: conda/install.py [hunk 2]
```diff
@@ -369,16 +369,18 @@ def replace_long_shebang(mode, data):
...
     return data
 
 
 def update_prefix(path, new_prefix, placeholder=prefix_placeholder, mode='text'):
-    if on_win and (placeholder != prefix_placeholder) and ('/' in placeholder):
-        # original prefix uses unix-style path separators
+    if on_win:
+        # force all prefix replacements to forward slashes to simplify need to escape backslashes
         # replace with unix-style path separators
         new_prefix = new_prefix.replace('\\', '/')
```
Comment (kalefranz) [inline conda/install.py]: At some point I'd like a full sweep/refactor that starts to make direct use of `os.path`.
e.g. https://github.com/python/cpython/blob/2.7/Lib/ntpath.py

Comment (msarahan) [inline conda/install.py]: OK, not sure what you mean here.  The purpose of this PR is to intentionally replace any backslashes with forward slashes, which sort of goes against how python deals with windows paths.  The purpose is that by doing this, you don't need to know what kind of file you're replacing prefixes in.  As it is, if we replace a prefix in a python file, we don't escape our backslashes, and the path gets misinterpreted.

File: conda/install.py [hunk 3]
```diff
@@ -369,16 +370,18 @@ def replace_long_shebang(mode, data):
 def replace_prefix(mode, data, placeholder, new_prefix):
     if mode == 'text':
         data = data.replace(placeholder.encode('utf-8'), new_prefix.encode('utf-8'))
-    elif mode == 'binary':
+    # Skip binary replacement in Windows.  Some files do have prefix information embedded, but
+    #    this should not matter, as it is not used for things like RPATH.
+    elif mode == 'binary' and on_win:
```
Comment (ilanschnell) [inline conda/install.py]: Shouldn't this be `not on_win`?

Comment (msarahan) [inline conda/install.py]: Good catch.  Fixed.

**Comment by kalefranz** [comment]
Legitimate error from appveyor:

    ____________________________ FileTests.test_binary ____________________________ 
    Traceback (most recent call last):
      File "C:\projects\conda\tests\test_install.py", line 111, in test_binary
        b'\x7fELF.../usr/local/lib/libfoo.so\0\0\0\0\0\0\0\0'
      File "C:\Python27_64\lib\unittest\case.py", line 513, in assertEqual
        assertion_func(first, second, msg=msg)
      File "C:\Python27_64\lib\unittest\case.py", line 506, in _baseAssertEqual
        raise self.failureException(msg)
    AssertionError: '\x7fELF.../some-placeholder/lib/libfoo.so\x00' != '\x7fELF.../usr/local/lib/libfoo.so\x00\x00\x00\x00\x00\x00\x00\x00'
     3 failed, 220 passed, 2 skipped, 2 xfailed, 1 pytest-warnings in 709.07 seconds  

https://ci.appveyor.com/project/ironmancio54716/conda/build/1.0.1370/job/a1kv6uew2yvpv4qj

**Comment by msarahan** [comment]
I'm not sure that that error is relevant, @kalefranz.  We don't do binary replacement on Windows.  This might be relevant for MSYS2 packages, though.  @mingwandroid - how should we be handling prefix replacement in binaries on Windows for relocatability?

**Comment by mingwandroid** [comment]
The `m2-` packages that use the POSIX emulation layer are all built with a prefix or `/usr`, and, no matter where the conda env is, the location of  msys-2.0.dll is reinterpreted as `/usr/bin`.

The`m2w64-` packages have been patched when necessary to implement dynamic relocation.

IOW `MSYS2` had always supported installing in different locations so we're fine.

**Comment by kalefranz** [comment]
So we at least need to fix the test then, right? Either skipif on_win, or do something to make the test pass?

**Comment by kalefranz** [comment]
P.S. re-queued a build on appveyor so we have the full output.
https://ci.appveyor.com/project/ironmancio54716/conda/build/1.0.1443

**Comment by msarahan** [comment]
@kalefranz all green here - not sure what the patch codecov is about.

**Pull request merged**